### DISCUSSION TITLE
`continue` label

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#5547] [Language Ideas]
## Prelude

In loops, the compiler creates two hidden labels, a `break` and a `continue` one. These special hidden labels are used when lowering `break;` and `continue;` statements into simpler `goto label;` ones, where `label` is the respective hidden label.

## Summary

Enable defining a label with the name `continue`. `continue;` statements, that are already equivalent to `goto continue;` when lowered, will execute code under the `continue` label.

## Motivation

While `continue` is a great tool to use in loops, it comes short when some additional operation has to be performed on *every* iteration, regardless of whether it got early terminated. Only `for` loops provide the ability to do it, but the syntax is wacky in multiple statements. As already mentioned in the prelude, `continue`, just like `break` are actual hidden labels that the compiler generates. Exposing the ability to use a `continue` label that the compiler will use gives the user better logic control and avoids yet another instance of using `goto` statements.

## Details

Only `for` loops contain iteration statements that are executed in the `continue` label. `while` and `do`-`while` loops do not contain any additional code. `foreach` acts pretty much like a `while` loop, but does not make use of the `continue` label itself, neither is there any code under it (I believe).

### Details about `foreach`
The `foreach` loop is transformed into a `while` loop like so:
```csharp
foreach (var value in values)
{
    // code
}

// ---

var enumerator = values.GetEnumerator();
while (enumerator.MoveNext())
{
    var value = enumerator.Current;
    // code
}
enumerator.Dispose();
```

Given the above code, it is evident that a `continue;` statement anywhere in the code within the `foreach` loop's body, would simply move to the next element in the enumerator.

The continuation code in a `for` loop is the code in the last statement (the "incrementor" statement). This may consist of multiple statements, separated by `,`. In this case, a user-defined `continue` label will include the incrementor statement, **after** all other statements.

The following example shows this:
```csharp
int sum = 0;
int odds = 0;
for (int i = 0; i < 10; i++)
{
    if (i % 2 is 0)
        continue;

    odds++;

continue:
    sum += i;
}
```

Will do the same as the currently valid code (but NOT the same):
```csharp
int sum = 0;
int odds = 0;
for (int i = 0; i < 10; i++)
{
    if (i % 2 is 0)
        goto _continue;

    odds++;

_continue:
    sum += i;
}
```

What this code is actually equivalent to is:
```csharp
int sum = 0;
int odds = 0;
{
    int i = 0;
    while (i < 10)
    {
        if (i % 2 is 0)
            continue;

        odds++;

    continue:
        sum += i;
        i++;
    }
}
```

**Comment by CyrusNajmabadi** [comment]
Why would you write:

```c#
int sum = 0;
int odds = 0;
for (int i = 0; i < 10; i++)
{
    if (i % 2 is 0)
        continue;

    odds++;

continue:
    sum += i;
}
```

instead of:

```c#
int sum = 0;
int odds = 0;
for (int i = 0; i < 10; sum += i, i++)
{
    if (i % 2 is 0)
        continue;

    odds++;
}
```

?

**Reply by Rekkonnect** [reply]
Obviously there's more complex scenarios than that. Imagine you have to use some function on an object to move its iterator along.

I have to currently write this code, which inspired this idea:
```csharp
var currentDirection = new DirectionalLocation(Direction.Right);
for (int i = 0; i < 4; i++, currentDirection.TurnRight())
{
    // code
}
```

P.S. Order of execution of "incrementor" statements is not guaranteed, and your equivalent code doesn't seem to be correct.

**Reply by CyrusNajmabadi** [reply]
> Obviously there's more complex scenarios than that. Imagine you have to use some function on an object to move its iterator along.

> I have to currently write this code, which inspired this idea:

That code seems fine to me.

> P.S. Order of execution of "incrementor" statements is not guaranteed, 

It is guaranteed: https://github.com/dotnet/csharplang/blob/main/spec/statements.md#the-for-statement.  Specifically:

> "the expressions of the for_iterator, if any, are evaluated in sequence,"

>  and your equivalent code doesn't seem to be correct.

Indeed, i did it as a quick rewrite and missed the data dependency.  I've revised it now.

**Reply by CyrusNajmabadi** [reply]
> Obviously there's more complex scenarios than that.

I'm not seeing why existing mechanisms are not suitable for those more complex scenarios.  We already have labeled goto.  So instead of needing something, why not actually just put all your incrementor logic into one place at teh end and just use a simple `goto incrementor;` statement.

**Reply by Rekkonnect** [reply]
> It is guaranteed: https://github.com/dotnet/csharplang/blob/main/spec/statements.md#the-for-statement. Specifically:
> 
> > "the expressions of the for_iterator, if any, are evaluated in sequence,"

Huh, look at that. I remember looking at the docs:tm: back in the day and seeing that there is no guarantee.

> That code seems fine to me.

Even if that's the case, other loops don't have that. This particular proposal doesn't empower `for` too much, only offers an additional way to express it. `while`, `do`-`while` and `foreach` still remain unable to serve this functionality without writing a custom label and using `goto incrementor;` instead of the friendlier `continue;` that has existed for this reason.

**Reply by 333fred** [reply]
> `while`, `do`-`while` and `foreach` still remain unable to serve this functionality without writing a custom label and using `goto incrementor;` instead of the friendlier `continue;` that has existed for this reason.

My main problem with such a proposal is that I think far too niche to spend time on. Of the loop constructs you mentioned, the only one I have ever wanted such functionality in is `foreach`, and extremely uncommonly. For scenarios where this is needed, the existing support of manually creating a label is nice and clear, requires no changes from the language, and only requires typing an extra few characters.

**Reply by CyrusNajmabadi** [reply]
>  and using goto incrementor;

Yes.  That seems fine to me.  It's a more complex scenario.  And there's a really good solution for it already IMO.  That solution is completely clear (IMO someone of practically any coding or C# level would understand it).  It's idiomatic, and it's been supported since C# 1.0.  It's a great solution without downsides AFAICT.  It's unclear to me why it would not be *preferred* over the solution above.

Indeed, i would be against making `continue` change meaning as it would mean all loops now became suspect.  I'm highly against features that mean one thing, unless you scan an arbitrary amoutn of code in the *future* to figure out if it might mean something different.
